Add EntryPointFilter option

sdmaclea opened the pull request:
Add an environment variable & property to add an entry point filter to log unhandled managed exceptions without depending on them being completely unhandled.

Part of https://github.com/dotnet/core-setup/issues/4836

sdmaclea committed:
Add EntryPointFilter

jkotas commented:
Build breaks...

jkotas commented:
> environment variable 

Do we actually expect this to be turned on via environment variable? I though we said that the host will set it as one of the key/value pairs.

sdmaclea commented:
> Build breaks...

Code is still WIP.  Looks like I will need a slightly different approach

> the host will set it as one of the key/value pairs.

that will ultimately be how this is done. @vitek-karas wanted a working prototype.  We were more concerned with whether it could be done then the details of how to enable it.

I haven't tested this yet, but given what I see it seems like it should be reasonable to debug.

vitek-karas commented:
Just curious - with this change, what happens from the point of view of the host? Can it still catch the exception (and avoid killing the process)?

sdmaclea commented:
The test I modified was disabled in #22889 for intermittent failures.

sdmaclea commented:
@sywhang I spent a bit of time on the test disabled by #22889.  

From what I could see the original test was parsing every application log entry.
I rearranged the code a bit to make it run faster/more stable.  
- Only look at the most recent log entries
- Only look at entries from '.NET Runtime'
- Removed timeout

I left the test disabled.

I tested it manually on my machine.

sdmaclea commented:
> Just curious - with this change, what happens from the point of view of the host? Can it still catch the exception (and avoid killing the process)?

It seems to work OK.  I am testing with a EX_TRY block.  It does still generate event log entries, but there are about a dozen `Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))` lines on the console.

vitek-karas commented:
@sdmaclea can you please confirm with @pakrym that the new behavior will work well for the ASP.NET scenario?

vitek-karas commented on src/inc/clrconfigvalues.h:
```diff
@@ -269,6 +269,7 @@ CONFIG_DWORD_INFO_DIRECT_ACCESS(INTERNAL_AssertOnFailFast, W("AssertOnFailFast")
 RETAIL_CONFIG_DWORD_INFO_EX(UNSUPPORTED_legacyCorruptedStateExceptionsPolicy, W("legacyCorruptedStateExceptionsPolicy"), 0, "Enabled Pre-V4 CSE behavior", CLRConfig::FavorConfigFile)
 CONFIG_DWORD_INFO_EX(INTERNAL_SuppressLostExceptionTypeAssert, W("SuppressLostExceptionTypeAssert"), 0, "", CLRConfig::REGUTIL_default)
 RETAIL_CONFIG_DWORD_INFO_EX(UNSUPPORTED_FailFastOnCorruptedStateException, W("FailFastOnCorruptedStateException"), 0, "Failfast if a CSE is encountered", CLRConfig::FavorConfigFile)
+RETAIL_CONFIG_DWORD_INFO_EX(INTERNAL_UseEntryPointFilter, W("UseEntryPointFilter"), 0, "", CLRConfig::REGUTIL_default)
```
Not sure we want to keep this way of enabling it... for testing you made corerun pass the runtime property, so this should not be strictly necessary anymore.

Using env. variable for this is troublesome since the variable is inherited by child processes and this behavior should be very host specific.

sdmaclea replied:
Done

vitek-karas reviewed (dismissed):
I think we should remove the COMPlus env. variable way of enabling this before merging.

sdmaclea commented:
/cc @elinor-fung

sdmaclea commented:
Here is the log from the manual test run.
```
D:\a\1\s\bin\tests\Windows_NT.x64.Debug\baseservices\exceptions\WindowsEventLog\WindowsEventLog>WindowsEventLog.cmd
BEGIN EXECUTION
 "D:\a\1\s\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe" WindowsEventLog.exe

Unhandled Exception: System.ArgumentException: PFO0IAKC81
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.UnhandledException(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 16
   at Program.Main(String[] args) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 185
Found 180 entries in Event Log
***      Event Log       ***
Application: corerun.exe
CoreCLR Version: 4.6.27506.0
Description: The process was terminated due to an unhandled exception.
Exception Info: System.ArgumentException: PFO0IAKC81
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.UnhandledException(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 16
   at Program.Main(String[] args) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 185

Checking for existence of : unhandled exception
Checking for existence of : ArgumentException
Checking for existence of : PFO0IAKC81
Found 180 entries in Event Log

Unhandled Exception: System.ArgumentException: 6QNPL0LXPB
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.UnhandledException(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 16
   at Program.Main(String[] args) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 185
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))
Found 181 entries in Event Log
***      Event Log       ***
Application: corerun.exe
CoreCLR Version: 4.6.27506.0
Description: The process was terminated due to an unhandled exception.
Exception Info: System.ArgumentException: 6QNPL0LXPB
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.UnhandledException(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 16
   at Program.Main(String[] args) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 185

Checking for existence of : unhandled exception
Checking for existence of : ArgumentException
Checking for existence of : 6QNPL0LXPB

Unhandled Exception: System.ArgumentException: YUJ4MWVBCC
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.UnhandledException(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 16
   at Program.Main(String[] args) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 185
Found 182 entries in Event Log
***      Event Log       ***
Application: corerun.exe
CoreCLR Version: 4.6.27506.0
Description: The process was terminated due to an unhandled exception.
Exception Info: System.ArgumentException: YUJ4MWVBCC
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.UnhandledException(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 16
   at Program.Main(String[] args) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 185

Checking for existence of : unhandled exception
Checking for existence of : ArgumentException
Checking for existence of : YUJ4MWVBCC
FailFast:
FHNDHAX47N
   at System.Environment.FailFast(System.String, System.Exception)
   at Program.FailFast(System.String)
   at Program.Main(System.String[])

Exception details:
System.ArgumentException: FHNDHAX47N
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.FailFast(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 26
Found 183 entries in Event Log
***      Event Log       ***
Application: corerun.exe
CoreCLR Version: 4.6.27506.0
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: FHNDHAX47N
Description: The process was terminated due to an unhandled exception.System.ArgumentException: FHNDHAX47N
   at Program.B(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 41
   at Program.A(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 36
   at Program.FailFast(String msg) in D:\a\1\s\tests\src\baseservices\exceptions\WindowsEventLog\WindowsEventLog.cs:line 26
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Program.FailFast(System.String)
   at Program.Main(System.String[])

Checking for existence of : The application requested process termination through System.Environment.FailFast(string message).
Checking for existence of : ArgumentException
Checking for existence of : FHNDHAX47N
Expected: 100
Actual: 100
END EXECUTION - PASSED
PASSED
```

jkotas commented on src/coreclr/hosts/corerun/corerun.cpp:
```diff
@@ -25,13 +26,29 @@ static const wchar_t *serverGcVar = W("COMPlus_gcServer");
 // On by default.
 static const wchar_t *concurrentGcVar = W("COMPlus_gcConcurrent");
 
+// Environment variable for setting whether to swallow uncaught exceptions (for testing)
```
We have links to CoreRun in the docs as reference implementations of the hosts: https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting#sample-hosts

What people do with this is that will copy&paste the corerun code and then keep going with it...

sdmaclea replied:
Hmmm…  I guess that means, I will switch back to environment variables for the test.

jkotas commented on src/coreclr/hosts/corerun/corerun.cpp:
```diff
@@ -25,13 +26,29 @@ static const wchar_t *serverGcVar = W("COMPlus_gcServer");
+// Environment variable for setting whether to swallow uncaught exceptions (for testing)
+static const wchar_t *swallowUncaughtExceptionsVar = W("COMPlus_Corerun_Swallow_Uncaught_Exceptions");
+static const wchar_t *useEntrypointFilterVar = W("COMPlus_Corerun_Use_EntryPoint_Filter");
+
```
There is a separate corerun for Unix. Do we need to test what this does on Unix too?

sdmaclea replied:
I'm not sure how to test on Linux.  @vitek-karas had suggested enabling this only for Windows since this was particularly for IIS.  I think I'll head that direction.

sdmaclea replied:
This is now disabled except on Windows

sdmaclea commented:
I removed the corerun changes, restored the environment variable, added a new mechanism and environment variable to simulate the host catching all the exceptions.

The test is passing and the anomalous `Internal CLR error. (0x80131506 (COR_E_EXECUTIONENGINE))` are no longer present.

jkotas commented on src/vm/corhost.cpp:
```diff
@@ -460,10 +460,23 @@ HRESULT CorHost2::ExecuteAssembly(DWORD dwAppDomainId,
+        if(CLRConfig::GetConfigValue(CLRConfig::INTERNAL_Corhost_Swallow_Uncaught_Exceptions))
         {
-            *pReturnValue = retval;
+            EX_TRY
```
These macros  are rightly coupled with the rest of CoreCLR. They do not simulate well what code outside CoreCLR.dll catching the exception.

sdmaclea replied:
They were at least able to suppress the event log which is triggered by the unhandled exception handler.  Which allowed testing the new mechanism to enable logging caught exceptions.

Do you have a different recommendation?  Creating a new host for this test seems like overkill....

sdmaclea replied:
I guess I could place it under `#ifdef PLATFORM_WINDOWS` and use `__try`/`__catch` if you think that would be better.

jkotas replied:
> Do you have a different recommendation?

Would it make sense to wait for https://github.com/dotnet/core-setup/pull/5336 ? Creating new hosts for testing should be simple with that, I think.

sdmaclea replied:
Given @pakrym & @davidfowl's priority on this, I am thinking we should merge as is, and open an issue to fix after https://github.com/dotnet/core-setup/pull/5336 is implemented.

I guess the alternative, would be to put the test in a separate PR and remove the environment variables from this PR.

jkotas replied:
I would be ok with not having tests included in the PR - if waiting for dotnet/core-setup#5336  makes sense.

sdmaclea replied:
I guess one other approach would be to 
+ keep the EX_TRY/EX_CATCH_HR, 
+ add another property/key pair (SWALLOW_UNCAUGHT_EXCEPTIONS) to allow it to be enabled
+ recommend ASP.NET's host use this mechanism.

sdmaclea replied:
@vitek-karas Do you have a preference?

vitek-karas replied:
I would only use the property as already implemented and merge that.
I can live with it not having tests for now.

That said, writing a one off test which manually initializes coreclr and invokes a super simple main should not be that hard (we can hardcode TPA and the rest of the needed properties). I know we don't really have hosting API tests, but that doesn't mean we should not try to add them :-)

AaronRobinsonMSFT commented on tests/src/baseservices/exceptions/WindowsEventLog/WindowsEventLog.cs:
```diff
@@ -144,7 +142,30 @@ public static int Main(string[] args)
+                return 1;
+            }
+
+            if (RunUnhandledExceptionTest(true))
```
Please use named arguments for literals.

sdmaclea commented:
test OSX10.12 x64 Checked Innerloop Build and Test
test Ubuntu x64 Formatting

sdmaclea commented:
I have removed the test.  I will create another PR for it, after I merge this.

jkotas commented on src/vm/excep.cpp:
```diff
@@ -5152,10 +5138,28 @@ LONG InternalUnhandledExceptionFilter(
 
 } // LONG InternalUnhandledExceptionFilter()
 
+bool& GetUseEntryPointFilter()
```
This is method looks weird. Can this be just a simple static variable and the ifdef moved to `ParseUseEntryPointFilter`?

sdmaclea replied:
Sure, but I thought you didn't like simple global static variables.  One of my earlier reviews there was concern with initialization startup races for global static variables.

jkotas replied:
We have hundreds of simple global static variables. I do not see a problem with them.

> concern with initialization startup races for global static variables

This must have been something slightly different. Maybe global static variables with destructors?

sdmaclea committed:
PR feedback

jkotas commented on src/vm/excep.cpp:
```diff
@@ -5152,10 +5138,20 @@ LONG InternalUnhandledExceptionFilter(
+
+void ParseUseEntryPointFilter(LPCWSTR value)
+{
+#ifndef PLATFORM_WINDOWS // This feature has only been tested on Windows, keep it disabled on other platforms
```
Should this be `#ifdef PLATFORM_WINDOWS` ?

sdmaclea replied:
Mornings

jkotas commented on src/vm/excep.cpp:
```diff
@@ -5152,10 +5138,20 @@ LONG InternalUnhandledExceptionFilter(
 
 } // LONG InternalUnhandledExceptionFilter()
 
+static bool useEntryPointFilter = false;
```
Nit: We typically prefix simple statics with `s_` prefix. (The codebase is not consistent though.)

jkotas commented on src/vm/excep.cpp:
```diff
@@ -5168,20 +5164,38 @@ LONG EntryPointFilter(PEXCEPTION_POINTERS pExceptionInfo, PVOID _pData)
+        return ret;
+    }
+
+    if(!useEntryPointFilter)
```
```suggestion
    if (!useEntryPointFilter)
```

jkotas commented on src/vm/excep.cpp:
```diff
@@ -5168,20 +5164,38 @@ LONG EntryPointFilter(PEXCEPTION_POINTERS pExceptionInfo, PVOID _pData)
         LOG((LF_EH, LL_INFO100, "EntryPointFilter: setting TSNC_ProcessedUnhandledException\n"));
         pThread->SetThreadStateNC(Thread::TSNC_ProcessedUnhandledException);
+
+        if(ret == EXCEPTION_EXECUTE_HANDLER)
```
```suggestion
        if (ret == EXCEPTION_EXECUTE_HANDLER)
```

jkotas reviewed (approved):
LGTM otherwise

davidfowl commented:
Just so I'm clear on what's being added here.
- We set the entrypoint filter option as an environment variable (at least for now since there's no way in hostfxr to pass host properties)
- We have code that runs and hooks AppDomain.UnhandledException that calls into our code to tell it an exception happened and to show a nice error page

Where I'm unclear:
- Do we still need to block in the UnhandledException callback to prevent crashes
- Do we still keep our try/catch in native code? 

I can't remember what we decided here.

sdmaclea commented:
test Ubuntu x64 Formatting
test CentOS7.1 x64 Debug Innerloop Build
test CentOS7.1 x64 Checked Innerloop Build and Test

sdmaclea commented:
@davidfowl 

> We set the entrypoint filter option as an environment variable

This doesn't have the environment variable yet. We were going to leave this dependent on the native hosting changes to add ability to set properties.

> We have code that runs and hooks AppDomain.UnhandledException that calls into our code to tell it an exception happened and to show a nice error page.

I have not tried to prototype this.  ASP.NET team were going to try this.  

> Do we still need to block in the UnhandledException callback to prevent crashes

No

> Do we still keep our try/catch in native code?

At the moment yes.  I had considered moving that into our code too so we could catch the exceptions in a coreclr friendly way.

davidfowl commented:
> This doesn't have the environment variable yet. We were going to leave this dependent on the native hosting changes to add ability to set properties.

Maybe leave in the env variable for now so we can test an end to end?

> I have not tried to prototype this. ASP.NET team were going to try this.

Yes we're going to do this.

> At the moment yes. I had considered moving that into our code too so we could catch the exceptions in a coreclr friendly way.

OK so even if we have the try catch things will get logged appropriately? The only change we have to prototype then is showing the nice error page on failure.

sdmaclea commented:
I am going to merge this because there is consensus, it is passing, and the environment variable had push back.  

I'll continue the discussion in the test PR (which is adding the environment variable already).

jkotas commented on src/vm/excep.cpp:
```diff
@@ -5152,10 +5138,20 @@ LONG InternalUnhandledExceptionFilter(
+
+void ParseUseEntryPointFilter(LPCWSTR value)
+{
+#ifdef PLATFORM_WINDOWS // This feature has only been tested on Windows, keep it disabled on other platforms
```
Is the code under the ifdef actually compiled on Windows? I do not think `PLATFORM_WINDOWS` is defined for the C/C++ anywhere today.

sdmaclea replied:
@jkotas This must be defined on Windows.  I put an #error inside the block to get it to fail compile to check for sure.  It is also used in a few other files in the VM eepolicy.cpp & peimagelayout.cpp.

jkotas replied:
Yes, it is fixed now. @elinor-fung added `PLATFORM_WINDOWS` to cmake scripts a few days ago.
